mds: introduce ceph.mirror.info vxattr

vshankar opened the pull request:
This virtual extended attribute is to be used when setting up
a filesystem for mirroring. The idea is to prevent multiple
ceph clusters adding a same remote ceph cluster as a peer for
a filesystem.

Adding a peer is configured via an interface exposed by manager
module ("mirroring" plugin). This extended attribute is set on
the remote filesystems root directory by the manager plugin
thereby preventing other clusters adding the filesystem as a
peer.

Note that this extended attribute is a constraint to be set only
on the root directory. Technically, filesystems in multiple
ceph clusters can mirror distinct directories to a filesystem
in the remote cluster. When supporting this, the path constraint
can be removed for setting the extended attribute on directories.

Signed-off-by: Venky Shankar <vshankar@redhat.com>

ukernel commented:
> This virtual extended attribute is to be used when setting up
> a filesystem for mirroring. The idea is to prevent multiple
> ceph clusters adding a same remote ceph cluster as a peer for
> a filesystem.
> 
> Adding a peer is configured via an interface exposed by manager
> module ("mirroring" plugin). This extended attribute is set on
> the remote filesystems root directory by the manager plugin
> thereby preventing other clusters adding the filesystem as a
> peer.
> 
> Note that this extended attribute is a constraint to be set only
> on the root directory. Technically, filesystems in multiple
> ceph clusters can mirror distinct directories to a filesystem
> in the remote cluster. When supporting this, the path constraint
> can be removed for setting the extended attribute on directories.
> 
> Signed-off-by: Venky Shankar [vshankar@redhat.com](mailto:vshankar@redhat.com)

no not just used normal user xattrs

vshankar commented:
> no not just used normal user xattrs

It needs to be in ceph.* namespace since this is used by ceph internal daemon.

batrick commented:
> > no not just used normal user xattrs
> 
> It needs to be in ceph.* namespace since this is used by ceph internal daemon.

I think Zheng's main objection here is that every inode will include the overhead of this `mirror_info` string, both on-disk and in-memory. Another thought: can we not store this `ceph.` namespace xattr in the xattr map? I haven't looked but it might just require teaching the MDS/client to allow some xattrs in the `ceph.` namespace to go in the xattrmap.

vshankar commented:
> I think Zheng's main objection here is that every inode will include the overhead of this `mirror_info` string, both on-disk and in-memory. Another thought: can we not store this `ceph.` namespace xattr in the xattr map? I haven't looked but it might just require teaching the MDS/client to allow some xattrs in the `ceph.` namespace to go in the xattrmap.

I did consider putting this in xattr_map -- the added benefit of that was we get XATTR_* semantics for free since this xattr would not be considered as virtual. However, that requires filtering this xattr in trace information sent to the client by the mds, which is kind of ok, I guess. But, we kind of break the asusmption that all ceph.* xattrs are "virtual".

ukernel commented:
why not storing this information in mdsmap?

vshankar commented:
> why not storing this information in mdsmap?

I initially planned to put this in FSMap since we already track things there: https://github.com/ceph/ceph/pull/36131. However, there is good chance that the mirror information would need to be per directory inode (right now, it is constrained to root) to support multiple clusters mirror distinct directory (snapshots) to another cluster.

ukernel commented:
> > why not storing this information in mdsmap?
> 
> I initially planned to put this in FSMap since we already track things there: #36131. However, there is good chance that the mirror information would need to be per directory inode (right now, it is constrained to root) to support multiple clusters mirror distinct directory (snapshots) to another cluster.

This is problematic. For example:

one user set mirror peer at dir /a/b
another user wants to set mirror at dir /a. The problem is no efficient way to check if any descendant dir has mirror peer.. 

I suggest adding a 'mirror peer -> path' map to mdsmap

vshankar commented:
> This is problematic. For example:
> 
> one user set mirror peer at dir /a/b
> another user wants to set mirror at dir /a. The problem is no efficient way to check if any descendant dir has mirror peer..

That's one of the reasons this xattr is constrained to be set only on root. One possibility would be to mark each path component with a xattr signifying that a subtree beneath is mirrored and the final path component with this xattr.

> 
> I suggest adding a 'mirror peer -> path' map to mdsmap

Two questions:
1. Which clusters mdsmap? Needs to be in remote clusters map, right?
2. How does this (mdsmap, fsmap, etc..) scale with storing potentially large number of this key/val pairs?

ukernel commented:
> 2\. How does this (mdsmap, fsmap, etc..) scale with storing potentially large number of this key/val pairs?

Ok. this is an issue

ukernel commented:
> > I think Zheng's main objection here is that every inode will include the overhead of this `mirror_info` string, both on-disk and in-memory. Another thought: can we not store this `ceph.` namespace xattr in the xattr map? I haven't looked but it might just require teaching the MDS/client to allow some xattrs in the `ceph.` namespace to go in the xattrmap.
> 
> I did consider putting this in xattr_map -- the added benefit of that was we get XATTR_* semantics for free since this xattr would not be considered as virtual. However, that requires filtering this xattr in trace information sent to the client by the mds, which is kind of ok, I guess. But, we kind of break the asusmption that all ceph.* xattrs are "virtual".

I checked the code. make getattr read ceph.* from xattr_map is easy. just needs a few line of code

vshankar commented:
> I checked the code. make getattr read ceph.* from xattr_map is easy. just needs a few line of code

Right -- my first implementation used xattr_map. I redid based on inode_t implementation as it looked like the base for implementing vxattrs. I'll push an update based on xattr_map.

batrick commented:
@vshankar ping me when this is ready please.

vshankar commented:
> @vshankar ping me when this is ready please.

please take a look.

This change allows xattr value to be split up in the server and store each split as a distinct entry in xattr_map. The client side changes reconstruct the xattr value from the splits.

vshankar commented:
@batrick @ukernel pinging for reviews.

batrick commented on src/mds/Server.h:
```diff
@@ -318,6 +319,113 @@ class Server {
+  void mirror_info_removexattr_handler(CInode *cur, InodeStoreBase::xattr_map_ptr xattrs,
+                                       const XattrOp &xattr_op);
+
+  bool is_ceph_vxattr(std::string_view xattr_name) {
```
This should be a private method.

vshankar replied:
ack

vshankar replied:
it's already private

batrick commented on src/mds/Server.h:
```diff
@@ -318,6 +319,113 @@ class Server {
+      return true;
+    }
+
+    return xattr_name == "ceph.mirror.info";
```
isn't this check redundant?

vshankar replied:
is_ceph_vxattr() does not check this xattr name -- those checks only include xattrs that do *not* get persisted in xattr_map.

batrick commented on src/mds/Server.h:
```diff
@@ -318,6 +319,113 @@ class Server {
+                            XattrOp *xattr_op);
+
+    // set xattr for an inode in xattr_map
+    void (Server::*setxattr)(CInode *cur, InodeStoreBase::xattr_map_ptr xattrs,
```
I'm not familiar with this syntax. Can you explain what the `Session::*` means?

vshankar replied:
The handler function should be a member function of Server class.

batrick commented on src/mds/Server.cc:
```diff
@@ -5932,25 +6113,29 @@ void Server::handle_client_removexattr(MDRequestRef& mdr)
-  if (pxattrs && pxattrs->count(mempool::mds_co::string(name)) == 0) {
-    dout(10) << "removexattr '" << name << "' and ENODATA on " << *cur << dendl;
-    respond_to_request(mdr, -ENODATA);
+  int r = (this->*(handler->validate))(cur, pxattrs, &xattr_op);
```
ditto, this syntax is also unknown to me.

vshankar replied:
https://isocpp.org/wiki/faq/pointers-to-members#dotstar-vs-arrowstar

(habit of using shorthand -- maybe just switch to using std::invoke)

batrick replied:
Yes, please do that instead.

vshankar commented:
jenkins test api

vshankar commented:
@batrick please take a look

batrick commented on src/mds/Server.cc:
```diff
@@ -5800,13 +5802,191 @@ void Server::handle_remove_vxattr(MDRequestRef& mdr, CInode *cur)
+                                    std::string &cluster_id, std::string &fs_id) {
+  dout(20) << "parsing name=" << name << ", value=" << value << dendl;
+
+  std::regex regex(Server::MirrorXattrInfo::MIRROR_INFO_REGEX);
```
```suggestion
  static const std::regex regex(Server::MirrorXattrInfo::MIRROR_INFO_REGEX);
```

I think would be valid?

batrick commented on src/client/Client.cc:
```diff
@@ -11797,8 +11797,12 @@ int Client::_listxattr(Inode *in, char *name, size_t size,
-  for (const auto& p : in->xattrs) {
-    size_t this_len = p.first.length() + 1;
+  for ([[maybe_unused]] const auto &[xattr_name, xattr_value_bl] : in->xattrs) {
+    if (xattr_name.compare(0, 5, "ceph.") == 0) {
```
```suggestion
    if (xattr_name.rfind("ceph.", 0) == 0) {
```

batrick commented on src/mds/Server.h:
```diff
@@ -318,6 +319,113 @@ class Server {
+
+  bool is_allowed_ceph_xattr(std::string_view xattr_name) {
+    // not a ceph xattr -- allow!
+    if (xattr_name.compare(0, 5, "ceph.") != 0) {
```
```suggestion
    if (xattr_name.rfind("ceph.", 0) != 0) {
```

batrick reviewed (changes requested):
just some nits, looks great to me

vshankar commented:
@batrick fixed and updated.

batrick commented on src/mds/Server.cc:
```diff
@@ -5821,13 +5823,191 @@ void Server::handle_remove_vxattr(MDRequestRef& mdr, CInode *cur)
+                           const std::string &xattr_name, int op, int flags) {
+  if (op == CEPH_MDS_OP_SETXATTR) {
+    if (xattrs) {
+      if ((flags & CEPH_XATTR_CREATE) && xattrs->count(mempool::mds_co::string(xattr_name))) {
```
```suggestion
      if ((flags & CEPH_XATTR_CREATE) && xattrs && xattrs->count(mempool::mds_co::string(xattr_name))) {
```

I think? Also for `RMXATTR` below. See these failures:

````
2020-10-07T05:41:15.102 INFO:tasks.ceph.mds.a.smithi198.stderr:*** Caught signal (Segmentation fault) **
2020-10-07T05:41:15.102 INFO:tasks.ceph.mds.a.smithi198.stderr: in thread 7f9c35f74700 thread_name:ms_dispatch
2020-10-07T05:41:15.102 INFO:tasks.ceph.mds.a.smithi198.stderr: ceph version 16.0.0-6209-gf0e17b88 (f0e17b8812dfdab73564df55d47fd1dafa530c61) pacific (dev)
2020-10-07T05:41:15.103 INFO:tasks.ceph.mds.a.smithi198.stderr: 1: /lib64/libpthread.so.0(+0x12dc0) [0x7f9c3d6f6dc0]
2020-10-07T05:41:15.103 INFO:tasks.ceph.mds.a.smithi198.stderr: 2: (std::_Rb_tree<std::__cxx11::basic_string<char, std::char_traits<char>, mempool::pool_allocator<(mempool::pool_index_t)26, char> >, std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, mempool::pool_allocator<(mempool::pool_index_t)26, char> > const, ceph::buffer::v15_2_0::ptr>, std::_Select1st<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, mempool::pool_allocator<(mempool::pool_index_t)26, char> > const, ceph::buffer::v15_2_0::ptr> >, std::less<std::__cxx11::basic_string<char, std::char_traits<char>, mempool::pool_allocator<(mempool::pool_index_t)26, char> > >, mempool::pool_allocator<(mempool::pool_index_t)26, std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, mempool::pool_allocator<(mempool::pool_index_t)26, char> > const, ceph::buffer::v15_2_0::ptr> > >::find(std::__cxx11::basic_string<char, std::char_traits<char>, mempool::pool_allocator<(mempool::pool_index_t)26, char> > const&) const+0x16) [0x55cbff60e346]
2020-10-07T05:41:15.103 INFO:tasks.ceph.mds.a.smithi198.stderr: 3: (Server::xattr_validate(CInode*, std::shared_ptr<std::map<std::__cxx11::basic_string<char, std::char_traits<char>, mempool::pool_allocator<(mempool::pool_index_t)26, char> >, ceph::buffer::v15_2_0::ptr, std::less<std::__cxx11::basic_string<char, std::char_traits<char>, mempool::pool_allocator<(mempool::pool_index_t)26, char> > >, mempool::pool_allocator<(mempool::pool_index_t)26, std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, mempool::pool_allocator<(mempool::pool_index_t)26, char> > const, ceph::buffer::v15_2_0::ptr> > > const>, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, int, int)+0x1c0) [0x55cbff5ca3d0]
2020-10-07T05:41:15.103 INFO:tasks.ceph.mds.a.smithi198.stderr: 4: (Server::default_xattr_validate(CInode*, std::shared_ptr<std::map<std::__cxx11::basic_string<char, std::char_traits<char>, mempool::pool_allocator<(mempool::pool_index_t)26, char> >, ceph::buffer::v15_2_0::ptr, std::less<std::__cxx11::basic_string<char, std::char_traits<char>, mempool::pool_allocator<(mempool::pool_index_t)26, char> > >, mempool::pool_allocator<(mempool::pool_index_t)26, std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, mempool::pool_allocator<(mempool::pool_index_t)26, char> > const, ceph::buffer::v15_2_0::ptr> > > const>, Server::XattrOp*)+0x51) [0x55cbff5caa51]
2020-10-07T05:41:15.103 INFO:tasks.ceph.mds.a.smithi198.stderr: 5: (Server::handle_client_removexattr(boost::intrusive_ptr<MDRequestImpl>&)+0x2eb) [0x55cbff5cc76b]
2020-10-07T05:41:15.103 INFO:tasks.ceph.mds.a.smithi198.stderr: 6: (Server::dispatch_client_request(boost::intrusive_ptr<MDRequestImpl>&)+0xcdb) [0x55cbff5efa7b]
2020-10-07T05:41:15.103 INFO:tasks.ceph.mds.a.smithi198.stderr: 7: (MDCache::dispatch_request(boost::intrusive_ptr<MDRequestImpl>&)+0x33) [0x55cbff695003]
2020-10-07T05:41:15.103 INFO:tasks.ceph.mds.a.smithi198.stderr: 8: (MDSContext::complete(int)+0x56) [0x55cbff858386]
2020-10-07T05:41:15.104 INFO:tasks.ceph.mds.a.smithi198.stderr: 9: (void finish_contexts<std::vector<MDSContext*, std::allocator<MDSContext*> > >(ceph::common::CephContext*, std::vector<MDSContext*, std::allocator<MDSContext*> >&, int)+0x8d) [0x55cbff557ead]
2020-10-07T05:41:15.104 INFO:tasks.ceph.mds.a.smithi198.stderr: 10: (Locker::eval(CInode*, int, bool)+0x4d8) [0x55cbff7338e8]
2020-10-07T05:41:15.104 INFO:tasks.ceph.mds.a.smithi198.stderr: 11: (Locker::handle_client_caps(boost::intrusive_ptr<MClientCaps const> const&)+0x2048) [0x55cbff743a38]
2020-10-07T05:41:15.104 INFO:tasks.ceph.mds.a.smithi198.stderr: 12: (Locker::dispatch(boost::intrusive_ptr<Message const> const&)+0x104) [0x55cbff745634]
2020-10-07T05:41:15.104 INFO:tasks.ceph.mds.a.smithi198.stderr: 13: (MDSRank::handle_message(boost::intrusive_ptr<Message const> const&)+0xa7c) [0x55cbff55e95c]
2020-10-07T05:41:15.104 INFO:tasks.ceph.mds.a.smithi198.stderr: 14: (MDSRank::_dispatch(boost::intrusive_ptr<Message const> const&, bool)+0x7bb) [0x55cbff560eeb]
2020-10-07T05:41:15.104 INFO:tasks.ceph.mds.a.smithi198.stderr: 15: (MDSRankDispatcher::ms_dispatch(boost::intrusive_ptr<Message const> const&)+0x55) [0x55cbff5614e5]
2020-10-07T05:41:15.105 INFO:tasks.ceph.mds.a.smithi198.stderr: 16: (MDSDaemon::ms_dispatch2(boost::intrusive_ptr<Message> const&)+0x128) [0x55cbff551758]
2020-10-07T05:41:15.105 INFO:tasks.ceph.mds.a.smithi198.stderr: 17: (DispatchQueue::entry()+0x126a) [0x7f9c3ed3148a]
2020-10-07T05:41:15.105 INFO:tasks.ceph.mds.a.smithi198.stderr: 18: (DispatchQueue::DispatchThread::entry()+0x11) [0x7f9c3ede27c1]
2020-10-07T05:41:15.105 INFO:tasks.ceph.mds.a.smithi198.stderr: 19: /lib64/libpthread.so.0(+0x82de) [0x7f9c3d6ec2de]
2020-10-07T05:41:15.105 INFO:tasks.ceph.mds.a.smithi198.stderr: 20: clone()
````

/ceph/teuthology-archive/pdonnell-2020-10-07_03:27:03-fs-wip-pdonnell-testing-20201007.015751-distro-basic-smithi/5503131/teuthology.log

and these other jobs

`5 jobs: ['5503131', '5502947', '5503039', '5503084', '5502993']`

vshankar replied:
That code block is already under `if (xattr) {... }`. However, the removexattr part does not have this check (the check existed before this refactor).

batrick replied:
Oh, you're right. It's likely the RMXATTR, yes.

batrick commented on src/mds/Server.h:
```diff
@@ -318,6 +319,113 @@ class Server {
+  void mirror_info_removexattr_handler(CInode *cur, InodeStoreBase::xattr_map_ptr xattrs,
+                                       const XattrOp &xattr_op);
+
+  bool is_ceph_vxattr(std::string_view xattr_name) {
```
```suggestion
  static bool is_ceph_vxattr(std::string_view xattr_name) {
```

batrick commented on src/mds/Server.h:
```diff
@@ -318,6 +319,113 @@ class Server {
+           xattr_name == "ceph.dir.pin.distributed"sv;
+  }
+
+  bool is_allowed_ceph_xattr(std::string_view xattr_name) {
```
```suggestion
  static bool is_allowed_ceph_xattr(std::string_view xattr_name) {
```

batrick commented on src/mds/Server.h:
```diff
@@ -318,6 +319,113 @@ class Server {
+  const XattrHandler* get_xattr_or_default_handler(std::string_view xattr_name);
+
+  // generic variant to set/remove xattr in/from xattr_map
+  int xattr_validate(CInode *cur, InodeStoreBase::xattr_map_const_ptr xattrs,
```
```suggestion
  int xattr_validate(CInode *cur, const InodeStoreBase::xattr_map_const_ptr& xattrs,
```

and the others

batrick reviewed (changes requested):
Please do a run through the `fs` suite with `--subset 3/4` when you've fixed the issues.

batrick commented:
and /ceph/teuthology-archive/pdonnell-2020-10-07_03:27:03-fs-wip-pdonnell-testing-20201007.015751-distro-basic-smithi/5502990/teuthology.log

vshankar commented:
> and /ceph/teuthology-archive/pdonnell-2020-10-07_03:27:03-fs-wip-pdonnell-testing-20201007.015751-distro-basic-smithi/5502990/teuthology.log

thx. I'll take a look.

vshankar commented:
fs suite run: https://pulpito.ceph.com/vshankar-2020-10-10_06:43:49-fs-wip-cephfs-mirror-vxattr-20201009-073149-distro-basic-smithi/

There are a couple of failures:

(Test failure: test_subvolume_retain_snapshot_invalid_recreate (tasks.cephfs.test_volumes.TestVolumes))
https://pulpito.ceph.com/vshankar-2020-10-10_06:43:49-fs-wip-cephfs-mirror-vxattr-20201009-073149-distro-basic-smithi/5512328/ -- seems unrelated.

(/home/ubuntu/cephtest/clone.client.0/qa/workunits/fs/snaps/untar_snap_rm.sh)
https://pulpito.ceph.com/vshankar-2020-10-10_06:43:49-fs-wip-cephfs-mirror-vxattr-20201009-073149-distro-basic-smithi/5512359/

```
2020-10-11T20:27:10.226 INFO:tasks.workunit.client.0.smithi096.stdout:removed 'linux-2.6.33/arch/powerpc/include/asm/rtc.h'
2020-10-11T20:27:10.226 INFO:tasks.workunit.client.0.smithi096.stdout:removed 'linux-2.6.33/arch/powerpc/include/asm/swab.h'
2020-10-11T20:27:10.227 INFO:tasks.workunit.client.0.smithi096.stdout:removed 'linux-2.6.33/arch/powerpc/include/asm/string.h'
2020-10-11T20:27:10.227 INFO:tasks.workunit.client.0.smithi096.stdout:removed 'linux-2.6.33/arch/powerpc/include/asm/ps3stor.h'
2020-10-11T20:27:10.329 INFO:tasks.daemonwatchdog.daemon_watchdog:MonitorThrasher failed
2020-10-11T20:27:10.329 INFO:tasks.daemonwatchdog.daemon_watchdog:BARK! unmounting mounts and killing all daemons
2020-10-11T20:27:10.330 INFO:tasks.cephfs.fuse_mount:Running fusermount -u on ubuntu@smithi096.front.sepia.ceph.com...
2020-10-11T20:27:10.330 INFO:teuthology.orchestra.run:Running command with timeout 1800
2020-10-11T20:27:10.330 INFO:teuthology.orchestra.run.smithi096:> (cd /home/ubuntu/cephtest && exec sudo fusermount -u /home/ubuntu/cephtest/mnt.0)
2020-10-11T20:27:10.355 INFO:tasks.workunit.client.0.smithi096.stdout:removed 'linux-2.6.33/arch/powerpc/include/asm/mmzone.h'
2020-10-11T20:27:10.355 INFO:tasks.workunit.client.0.smithi096.stdout:removed 'linux-2.6.33/arch/powerpc/include/asm/immap_86xx.h'
2020-10-11T20:27:10.355 INFO:tasks.workunit.client.0.smithi096.stdout:removed 'linux-2.6.33/arch/powerpc/include/asm/processor.h'
2020-10-11T20:27:10.355 INFO:tasks.workunit.client.0.smithi096.stdout:removed 'linux-2.6.33/arch/powerpc/include/asm/qe.h'
2020-10-11T20:27:10.355 INFO:tasks.workunit.client.0.smithi096.stdout:removed 'linux-2.6.33/arch/powerpc/include/asm/hydra.h'
2020-10-11T20:27:10.356 INFO:tasks.workunit.client.0.smithi096.stdout:removed 'linux-2.6.33/arch/powerpc/include/asm/cpm2.h'
2020-10-11T20:27:10.356 INFO:tasks.workunit.client.0.smithi096.stdout:removed 'linux-2.6.33/arch/powerpc/include/asm/lppaca.h'
2020-10-11T20:27:10.356 INFO:tasks.workunit.client.0.smithi096.stdout:removed 'linux-2.6.33/arch/powerpc/include/asm/termbits.h'
2020-10-11T20:27:10.356 INFO:tasks.workunit.client.0.smithi096.stdout:removed 'linux-2.6.33/arch/powerpc/include/asm/dma.h'
2020-10-11T20:27:10.356 INFO:tasks.workunit.client.0.smithi096.stdout:removed 'linux-2.6.33/arch/powerpc/include/asm/byteorder.h'
2020-10-11T20:27:10.356 INFO:tasks.workunit.client.0.smithi096.stdout:removed 'linux-2.6.33/arch/powerpc/include/asm/sparsemem.h'
2020-10-11T20:27:10.356 INFO:tasks.workunit.client.0.smithi096.stdout:removed 'linux-2.6.33/arch/powerpc/include/asm/spu_csa.h'
2020-10-11T20:27:10.357 INFO:tasks.workunit.client.0.smithi096.stdout:removed 'linux-2.6.33/arch/powerpc/include/asm/types.h'
2020-10-11T20:27:10.357 INFO:tasks.workunit.client.0.smithi096.stdout:removed 'linux-2.6.33/arch/powerpc/include/asm/setup.h'
2020-10-11T20:27:10.357 INFO:tasks.workunit.client.0.smithi096.stdout:removed 'linux-2.6.33/arch/powerpc/include/asm/mpc52xx.h'
2020-10-11T20:27:10.357 INFO:tasks.workunit.client.0.smithi096.stderr:rm: cannot remove 'linux-2.6.33/arch/powerpc/include/asm/irq.h': Software caused connection abort
2020-10-11T20:27:10.357 INFO:tasks.workunit.client.0.smithi096.stderr:rm: cannot remove 'linux-2.6.33/arch/powerpc/include/asm/elf.h': Transport endpoint is not connected
2020-10-11T20:27:10.357 INFO:tasks.workunit.client.0.smithi096.stderr:rm: cannot remove 'linux-2.6.33/arch/powerpc/include/asm/pgtable-ppc64-64k.h': Transport endpoint is not connected
...
...
...
2020-10-11T20:27:10.381 INFO:tasks.workunit.client.0.smithi096.stderr:rm: cannot remove 'linux-2.6.33/arch/ia64': Transport endpoint is not connected
2020-10-11T20:27:10.381 INFO:tasks.workunit.client.0.smithi096.stderr:rm: cannot remove 'linux-2.6.33/arch/frv': Transport endpoint is not connected
2020-10-11T20:27:10.382 INFO:tasks.workunit.client.0.smithi096.stderr:rm: fts_read failed: Transport endpoint is not connected
2020-10-11T20:27:10.382 INFO:teuthology.orchestra.run.smithi096.stderr:fusermount: failed to unmount /home/ubuntu/cephtest/mnt.0: Device or resource busy
2020-10-11T20:27:10.382 DEBUG:teuthology.orchestra.run:got remote process result: 1
2020-10-11T20:27:10.382 INFO:tasks.workunit:Stopping ['fs/snaps'] on client.0...
```
This one too looks unrelated.

FYI, I have included the fixes as separate commits. Will squash those to the appropriate commits once you have done reviewing. Thx.

batrick commented:
Those failures are known and unrelated. Please go ahead and squash. I'll do another round of QA in multimds once you're done.

vshankar commented:
@batrick done

batrick commented:
https://pulpito.ceph.com/?branch=wip-pdonnell-testing-20201013.172209

failures known.
